patches/server/1059-More-Raid-API.patch

```diff
@@ -0,0 +1,106 @@
+@@ -149,6 +154,11 @@ public class Raid {
+                 this.heroesOfTheVillage.add(NbtUtils.loadUUID(nbtbase));
+             }
+         }
++        // Paper start
++        if (nbt.contains(PDC_NBT_KEY, net.minecraft.nbt.Tag.TAG_COMPOUND)) {
++            this.persistentDataContainer.putAll(nbt.getCompound(PDC_NBT_KEY));
++        }
++        // Paper end
+ 
+     }
+ 
+@@ -859,6 +869,11 @@ public class Raid {
+         }
+ 
+         nbt.put("HeroesOfTheVillage", nbttaglist);
++        // Paper start
++        if (!this.persistentDataContainer.isEmpty()) {
++            nbt.put(PDC_NBT_KEY, this.persistentDataContainer.toTagCompound());
++        }
++        // Paper end
+         return nbt;
+     }
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
+index 9650859b40576ff1d03b77ab3594499e47eeaf6a..611fc24ee90644359010b8d2bfb90f9eb2a04675 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
+@@ -103,4 +103,34 @@ public final class CraftRaid implements Raid {
+     public net.minecraft.world.entity.raid.Raid getHandle() {
+         return this.handle;
+     }
++
++    // Paper start - more Raid API
++    @Override
++    public int getId() {
++        return this.handle.getId();
++    }
++
++    @Override
++    public org.bukkit.boss.BossBar getBossBar() {
++        return new org.bukkit.craftbukkit.boss.CraftBossBar(this.handle.raidEvent);
++    }
++
++    @Override
++    public org.bukkit.persistence.PersistentDataContainer getPersistentDataContainer() {
++        return this.handle.persistentDataContainer;
++    }
++
++    @Override
++    public boolean equals(final Object o) {
++        if (this == o) return true;
++        if (o == null || this.getClass() != o.getClass()) return false;
++        final org.bukkit.craftbukkit.CraftRaid craftRaid = (org.bukkit.craftbukkit.CraftRaid) o;
++        return this.handle.equals(craftRaid.handle);
++    }
++
++    @Override
++    public int hashCode() {
++        return this.handle.hashCode();
++    }
++    // Paper end - more Raid API
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+index 01797d9791f19dfda4b168218eadeaae97f11eab..00357d78182b3ff87e3d9a45705b072af56739c8 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+@@ -2397,6 +2397,14 @@ public class CraftWorld extends CraftRegionAccessor implements World {
+         return (raid == null) ? null : new CraftRaid(raid);
+     }
+ 
++    // Paper start - more Raid API
++    @Override
++    public @Nullable Raid getRaid(final int id) {
++        final net.minecraft.world.entity.raid.@Nullable Raid nmsRaid = this.world.getRaids().raidMap.get(id);
```

Comment:
You could use the regular getter here instead of the internal field

Reply:
The getter isn't annotated correctly. It thinks it returns a nonnull value which isn't true since its just a Map#get.